### DISCUSSION TITLE
Model relations validation in full_clean()

### DISCUSSION THREAD

**Discussion by ralnss** [wagtail/wagtail#12423] [Q&A]
Greeting fellow wagtailians,

I already asked about this on [stackoverflow](https://stackoverflow.com/questions/79066878/wagtail-how-to-validate-page-model-relations-before-saving).
I have a page type where I want to define the title (and slug) automatically, based on some other fields and higher level model relations, mainly a foreign key and a relation based on a `MultipleChooserPanel`. To achieve that, I override the full_clean() method of my page class. So far so good.
What I want to achieve is basically:

```
def full_clean(self, *args, **kwargs):
    if not_all_relevant_model_data_available_yet:
        return super().full_clean(*args, **kwargs)

    # in this two functions I access the model stuff I need
    self.title = self.set_title()
    self.slug = self.set_slug()

    if not is_unique_among_siblings(self.slug):
        raise ValidationError()
    
    return super().full_clean(*args, **kwargs)
```
Long story short: In the call of `full_clean()`, in which all the model relations I need are resolved, raising the `ValidationError` does not result in a nice info banner in the gui, instead I get the traceback. Am I doing something wrong? Should model relations involving MultipleChooserPanels be handled differently?

I noticed that `full_clean()` is called again after saving, too, leading to more complications about this, since after saving a valid page, it's slug can not be unique among its siblings any more and the ValidationError would be raised unneccessarily. That's why I currently just catch and re-raise the ValidationError from the save action if the slug isn't unique.

I was thinking about using django constraints, but since these work on database level and not all of the needed data for validation are fields, it's not applicable here.

Why is it that ValidationErrors raised by `full_clean()` are sometimes shown in the gui nicely and sometimes not? Is this intended? Is there some solution that I'm missing? Should I just live with it and show some error template?

Any help appreciated

**Comment by Ganesh-403** [comment]
@ralnss I think the key issue is *where* the validation is happening rather than the `MultipleChooserPanel` itself.

`full_clean()` is primarily intended to validate the model's state. During page editing, Wagtail also has to deal with unsaved child objects (such as those managed by `InlinePanel`/`MultipleChooserPanel`), and the timing of when those relations are available differs throughout the edit/save lifecycle. :contentReference[oaicite:0]{index=0}

A couple of observations from your example:

1. **Overriding `clean()` is generally a better fit than `full_clean()`**

   Django's `full_clean()` orchestrates validation (`clean_fields()`, `clean()`, uniqueness checks, etc.). Custom validation logic is typically placed in `clean()`, leaving `full_clean()` itself alone.

   For example:

   ```python
   from django.core.exceptions import ValidationError

   class MyPage(Page):
       ...

       def clean(self):
           super().clean()

           if not self.pk:
               # or another condition if related objects aren't available yet
               return

           self.title = self.set_title()
           self.slug = self.set_slug()

           if not self.is_unique_among_siblings():
               raise ValidationError({
                   "slug": "A sibling page with this slug already exists."
               })
   ```

2. **Avoid validating against the current instance**

   After the page has been saved once, a uniqueness query such as

   ```python
   Page.objects.filter(
       path__startswith=self.get_parent().path,
       slug=self.slug,
   )
   ```

   should normally exclude the current page:

   ```python
   qs = Page.objects.sibling_of(self).filter(slug=self.slug)

   if self.pk:
       qs = qs.exclude(pk=self.pk)

   if qs.exists():
       ...
   ```

   Otherwise you'll always find yourself after the initial save.

3. **Why do you get a traceback instead of a banner?**

   From what I can tell, Wagtail expects validation errors raised during the normal form validation process to be attached to the form/model. If a `ValidationError` is raised later in the save lifecycle (for example after form validation has already completed), it may propagate as an exception instead of being rendered as a validation message. I couldn't find documentation stating that overriding `full_clean()` to raise validation errors after related objects have been processed is a supported extension point. :contentReference[oaicite:1]{index=1}

One question that would help narrow this down:

- **Which Wagtail version are you using?**

Also, when the traceback occurs, is it raised:
- while clicking **Save Draft**,
- **Publish**,
- or only after the page has already been written to the database?

That would help determine whether this is an issue with the validation timing or whether you're hitting a specific part of Wagtail's page save workflow.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
